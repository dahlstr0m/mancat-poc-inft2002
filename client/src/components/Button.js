// @flow

import * as React from 'react';
import { Component } from 'react-simplified';

/**
 * Renders a success button using Bootstrap styles.
 */
class ButtonSuccess extends Component<{
  onClick: () => mixed,
  small?: boolean,
  children?: React.Node,
}> {
  render() {
    return (
      <button
        type="button"
        className={'btn btn-success' + (this.props.small ? ' btn-sm py-0' : '')}
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a danger button using Bootstrap styles.
 */
class ButtonDanger extends Component<{
  onClick: () => mixed,
  small?: boolean,
  children?: React.Node,
}> {
  render() {
    return (
      <button
        type="button"
        className={'btn btn-danger' + (this.props.small ? ' btn-sm py-0' : '')}
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a light button using Bootstrap styles.
 */
class ButtonLight extends Component<{
  onClick: () => mixed,
  small?: boolean,
  children?: React.Node,
}> {
  render() {
    return (
      <button
        type="button"
        className={'btn btn-light' + (this.props.small ? ' btn-sm py-0' : '')}
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a button using Bootstrap styles.
 */
export default class Button {
  static Success = ButtonSuccess;
  static Danger = ButtonDanger;
  static Light = ButtonLight;
}
