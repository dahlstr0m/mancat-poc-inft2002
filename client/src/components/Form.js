// @flow

import * as React from 'react';
import { Component } from 'react-simplified';

/**
 * Renders a form label using Bootstrap styles.
 */
class FormLabel extends Component<{ children?: React.Node }> {
  render() {
    return <label className="col-form-label">{this.props.children}</label>;
  }
}

/**
 * Renders a form input using Bootstrap styles.
 */
class FormInput extends Component<{
  type: string,
  value: React.Node,
  onChange: (SyntheticEvent<HTMLInputElement>) => mixed,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, width, height, pattern
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { type, value, onChange, ...rest } = this.props;
    return (
      <input
        {...rest}
        className="form-control"
        type={this.props.type}
        value={this.props.value}
        onChange={this.props.onChange}
      />
    );
  }
}

/**
 * Renders a form textarea using Bootstrap styles.
 */
class FormTextarea extends React.Component<{
  value: React.Node,
  onChange: (SyntheticEvent<HTMLTextAreaElement>) => mixed,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, rows, cols
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { value, onChange, ...rest } = this.props;
    return <textarea {...rest} className="form-control" value={value} onChange={onChange} />;
  }
}

/**
 * Renders a form checkbox using Bootstrap styles.
 */
class FormCheckbox extends Component<{
  checked: React.Node,
  onChange: (SyntheticEvent<HTMLInputElement>) => mixed,
  labeling: React.Node,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, width, height, pattern
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { labeling, checked, onChange, ...rest } = this.props;
    return (
      <div className="custom-control custom-checkbox">
        <input
          {...rest}
          type="checkbox"
          className="custom-control-input"
          id="customCheck"
          checked={checked}
          onChange={onChange}
        />
        <label className="custom-control-label" htmlFor="customCheck">
          {labeling}
        </label>
      </div>
    );
  }
}

/**
 * Renders a form select using Bootstrap styles.
 */
class FormSelect extends Component<{
  value: React.Node,
  onChange: (SyntheticEvent<HTMLSelectElement>) => mixed,
  children?: React.Node,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, size.
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { value, onChange, children, ...rest } = this.props;
    return (
      <select {...rest} className="custom-select" value={value} onChange={onChange}>
        {children}
      </select>
    );
  }
}

/**
 * Renders a form date using Bootstrap styles.
 */
class FormDate extends Component<{
  type: string,
  value: React.Node,
  onChange: (SyntheticEvent<HTMLInputElement>) => mixed,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, width, height, pattern
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { type, value, onChange, ...rest } = this.props;
    return (
      <input
        {...rest}
        className="form-control"
        type={this.props.type}
        value={this.props.value}
        onChange={this.props.onChange}
      />
    );
  }
}

/**
 * Renders form components using Bootstrap styles.
 */
export default class Form {
  static Label = FormLabel;
  static Input = FormInput;
  static Textarea = FormTextarea;
  static Checkbox = FormCheckbox;
  static Select = FormSelect;
  static Date = FormDate;
}
