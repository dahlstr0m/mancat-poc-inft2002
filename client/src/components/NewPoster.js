// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import { history } from '../index';
import Button from './Button';
import Form from './Form';
import { Card, CardPlain } from './Card';
import { Alert } from './Widgets';
import { projectService, posterService } from '../services';

/**
 * New Poster page
 */
export default class NewPoster extends Component {
  poster: Poster = {
    // Default values for poster, updated on change in form
    posterId: 0,
    projectId: 1,
    posterDescription: '',
    posterUrl: '',
    thumbnailUrl: '',
  };
  projects: Project[] = [];
  posters: Poster[] = [];

  render() {
    return (
      <>
        <Card title="Add a new poster">
          <Form.Label> Connect to project:</Form.Label>
          <Form.Select
            value={this.poster.projectId}
            onChange={(event) => (this.poster.projectId = parseInt(event.currentTarget.value))}
          >
            {this.projects.map((project) => (
              <option key={project.projectId} value={project.projectId}>
                {project.title}
              </option>
            ))}
          </Form.Select>
          <br />
          <CardPlain>
            <Form.Label> Poster description:</Form.Label>
            <Form.Input
              type="text"
              value={this.poster.posterDescription}
              onChange={(event) => (this.poster.posterDescription = event.currentTarget.value)}
            ></Form.Input>
            <br />
            <Form.Label> Url to Poster:</Form.Label>
            <Form.Input
              type="text"
              value={this.poster.posterUrl}
              onChange={(event) => (this.poster.posterUrl = event.currentTarget.value)}
            ></Form.Input>
            <br />
            <Form.Label> Url to posters thumbnail:</Form.Label>
            <Form.Input
              type="text"
              value={this.poster.thumbnailUrl}
              onChange={(event) => (this.poster.thumbnailUrl = event.currentTarget.value)}
            ></Form.Input>
            <br />
          </CardPlain>
          <Button.Success
            onClick={() =>
              posterService
                .createPoster(this.poster)
                .then(() => {
                  history.push('/admin');
                  Alert.success('Poster added successfully');
                })
                .catch((error: Error) => Alert.danger('Error storing new poster: ' + error))
            }
          >
            Add new poster
          </Button.Success>
        </Card>
        <Card title="These posters already exists for this project:">
          {this.posters
            .filter((p) => p.projectId == this.poster.projectId)
            .map((p) => (
              <img
                key={p.posterId}
                src={p.thumbnailUrl}
                alt={'Can not display poster with id ' + p.posterId}
                width="105"
                heigth="150"
              />
            ))}
        </Card>
      </>
    );
  }
  mounted() {
    projectService
      .getProjects()
      .then((projects) => (this.projects = projects))
      .catch((error: Error) => Alert.danger('Error getting employers: ' + error));
    posterService
      .getPosters()
      .then((posters) => (this.posters = posters))
      .catch((error: Error) => Alert.danger('Error fetching connected posters: ' + error));
  }
}
