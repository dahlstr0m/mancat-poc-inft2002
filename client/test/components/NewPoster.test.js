// @flow

import * as React from 'react';
import { shallow } from 'enzyme';

// Components
import NewPoster from '../../src/components/NewPoster';
import Form from '../../src/components/Form';
import Button from '../../src/components/Button';

// Mocks
jest.mock('../../src/services/project-service');
jest.mock('../../src/services/poster-service');

// Tests
describe('NewPoster component tests', () => {
  test('NewPoster draws correctly', (done) => {
    const wrapper = shallow(<NewPoster />);

    // Wait for events to complete
    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();

      done();
    });
  });

  test('NewPoster adds poster correctly', (done) => {
    const wrapper = shallow(<NewPoster />);

    wrapper.find(Form.Select).simulate('change', { currentTarget: { value: 2 } });
    // $FlowExpectedError
    expect(wrapper.containsMatchingElement(<Form.Select value={2} />)).toEqual(true);

    wrapper
      .find(Form.Input)
      .at(0)
      .simulate('change', { currentTarget: { value: 'Description' } });
    wrapper
      .find(Form.Input)
      .at(1)
      .simulate('change', { currentTarget: { value: 'URL' } });
    wrapper
      .find(Form.Input)
      .at(2)
      .simulate('change', { currentTarget: { value: 'Thumb URL' } });

    // $FlowExpectedError
    expect(wrapper.containsMatchingElement(<Form.Input value="Description" />)).toEqual(true);
    // $FlowExpectedError
    expect(wrapper.containsMatchingElement(<Form.Input value="URL" />)).toEqual(true);
    // $FlowExpectedError
    expect(wrapper.containsMatchingElement(<Form.Input value="Thumb URL" />)).toEqual(true);

    wrapper.find(Button.Success).simulate('click');

    setTimeout(() => {
      expect(location.hash).toEqual('#/admin');
      done();
    });
  });
});
