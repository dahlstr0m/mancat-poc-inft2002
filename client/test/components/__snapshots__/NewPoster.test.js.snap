// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewPoster component tests NewPoster draws correctly 1`] = `
<Fragment>
  <Card
    title="Add a new poster"
  >
    <FormLabel>
       Connect to project:
    </FormLabel>
    <FormSelect
      onChange={[Function]}
      value={1}
    >
      <option
        key="1"
        value={1}
      >
        Project 1
      </option>
      <option
        key="2"
        value={2}
      >
        Project 2
      </option>
      <option
        key="3"
        value={3}
      >
        Project 3
      </option>
    </FormSelect>
    <br />
    <Card>
      <FormLabel>
         Poster description:
      </FormLabel>
      <FormInput
        onChange={[Function]}
        type="text"
        value=""
      />
      <br />
      <FormLabel>
         Url to Poster:
      </FormLabel>
      <FormInput
        onChange={[Function]}
        type="text"
        value=""
      />
      <br />
      <FormLabel>
         Url to posters thumbnail:
      </FormLabel>
      <FormInput
        onChange={[Function]}
        type="text"
        value=""
      />
      <br />
    </Card>
    <ButtonSuccess
      onClick={[Function]}
    >
      Add new poster
    </ButtonSuccess>
  </Card>
  <Card
    title="These posters already exists for this project:"
  >
    <img
      alt="Can not display poster with id 1"
      heigth="150"
      key="1"
      src="Poster 1 thumb url"
      width="105"
    />
  </Card>
</Fragment>
`;
